domain/application/service/unit.go

```diff
@@ -560,3 +562,62 @@ func (s *Service) GetUnitWorkloadVersion(ctx context.Context, unitName coreunit.
 	}
 	return version, nil
 }
+
+// GetUnitPublicAddress returns the public address for the specified unit.
+// For k8s provider, it will return the first public address of the cloud
+// service if any, the first public address of the cloud container otherwise.
+// For machines provider, it will return the first public address of the
+// machine.
+//
+// The following errors may be returned:
+// - [uniterrors.UnitNotFound] if the unit does not exist
+// - [network.NoAddressError] if the unit has no public address associated
+func (s *Service) GetUnitPublicAddress(ctx context.Context, unitName coreunit.Name) (network.SpaceAddress, error) {
+	unitUUID, err := s.st.GetUnitUUIDByName(ctx, unitName)
+	if err != nil {
+		return network.SpaceAddress{}, errors.Capture(err)
+	}
+	addrs, err := s.st.GetUnitAddresses(ctx, unitUUID)
+	if err != nil {
+		return network.SpaceAddress{}, errors.Capture(err)
+	}
+
+	// First match the scope, then sort by origin.
+	matchedAddrs := addrs.AllMatchingScope(network.ScopeMatchPublic)
```

Comment:
This looks like it replicates the current logic for setting the preferred public address for machine-resident units.

However, there's a slight deviation for k8s units. The old state logic has _strict_ public scope matching, which means we return a no-address error if there is no public address, without falling back as for machines.

Let's go with this though. We can evolve as required.